crates/cext/src/dag.rs

```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
         }
     }
 }
+/// @ingroup QkDag
+/// Return the operation nodes in the DAG listed in topological order.
+///
+/// @param dag A pointer to the DAG.
+/// @param order A pointer to an array of ``qk_dag_num_op_nodes(dag)`` elements
+/// of type ``uint32_t``, where this function will write the output to.
+///
+/// # Example
+/// ```c
+/// QkDag *dag = qk_dag_new();
+/// QkQuantumRegister *qr = qk_quantum_register_new(1, "my_register");
+/// qk_dag_add_quantum_register(dag, qr);
+///
+/// uint32_t qubit[1] = {0};
+/// qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+/// qk_dag_apply_gate(dag, QkGate_S, qubit, NULL, false);
+///
+/// // get the number of operation nodes
+/// uint32_t num_ops = qk_dag_num_op_nodes(dag); // 2
+/// uint32_t *order = malloc(sizeof(uint32_t) * num_ops);
+///
+/// // get operation nodes listed in topological order
+/// qk_dag_topological_op_nodes(dag, order);
+///
+/// // do something with the ordered nodes
+/// for (uint32_t i=0; i<num_ops; i++) {
+///     QkGate gate = qk_dag_op_node_gate_op(dag, order[i], NULL);
+///     printf("The gate at location %u is %u.\n", i, gate);
+/// }
+///
+/// // free the order array
+/// free(order);
+/// qk_quantum_register_free(qr);
+/// qk_dag_free(dag);
+/// ```
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``
+/// or if ``order`` is not a valid, non-null pointer to a sequence of ``qk_dag_num_op_nodes(dag)``
+/// consecutive elements of ``uint32_t``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, order: *mut u32) {
```

Comment:
```suggestion
pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out_order: *mut u32) {
```
I think it's nicer to indicate by name that this is an output parameter.

Reply:
It's fairly normal in C APIs that pointers to `const` are implicitly "in" params and pointers to non-`const` are implicitly out. Either way, I don't think we've used an `out_` prefix in other cases we do this (e.g. the transpile result).

Reply:
The `const` vs non-`const` convention makes sense, but for the non-`const` cases we already have exceptions in the C API in the form of in-place modifies, like the standalone passes which change the input circuit, or the layout configuration setters, or the `*_free` functions which don't return a value at all. These cases are not the typical populate-user-buffers pattern. 
The `out_` convention is already used BTW in `qk_dag_op_node_gate_op`. So, whatever we decide to land on - we should make the API consistent in this regard. Let's decide on a convention so we can use it consistently.

Reply:
I didn't mention in-place, but that's still kind of part of the same conventions.  Really though, there's no one-size-fits-all, and in my experience, I've not seen C APIs that tried to do that.

I'm not really _against_ calling it `out_order`, I just think it's unnecessary and so wordier than it could be. I think there's somewhat a false split being made - as you mentioned with mutate-in-place parameters, "in" and "out" params aren't mutually exclusive. The `mut` alone makes it clear we'll be writing to it in some form, at which point we have to check the docs.

If you think it'd be much clearer as `out_order` that's ok by me.

Reply:
I have renamed `order` to `out_order` as per review comments (almost typed `out_of_order` instead).